doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+=================================================
+NEP 41 — First step towards a new Datatype System
+=================================================
+
+:title: Improved Datatype Support
+:Author: Sebastian Berg
+:Author: Add...
+:Status: Draft
+:Type: Standard Track
+:Created: 2020-02-03
+
+
+.. note::
+
+    This NEP is part of a series of NEPs encompassing first information
+    about the previous dtype implementation and issues with it in NEP 40.
+    NEP 41 (this document) then provides an overview and generic design
+    choices for the refactor.  
+    Further NEPs 42 and 43 go into the technical dtypes of the datatype
+    and universal function related internal and external API changes.
+    In some cases it may be necessary to consult the other NEPs for a full
+    picture of the desired changes and why these changes are necessary.
+
+
+Abstract
+--------
+
+`Datatypes <data-type-objects-dtype>` in NumPy describe how to interpret each
+element in arrays. NumPy provides ``int``, ``float``, and ``complex`` numerical
+types, as well as string, datetime, and structured (or `~numpy.recarray`)
+datatype capabilities.  The growing Python community, however, has need for
+more diverse datatypes.
+Examples are datatypes with unit information attached (such as meters) or
+categorical datatypes (fixed set of possible values).
+However, the current NumPy datatype API is too limited to allow the creation
+of these.
+
+This NEP is the first step to enable such growth; it will lead to
+a simpler development path for new datatypes. Support is also proposed to
+create datatypes from Python rather than in C.
+By refactoring our datatype API and improving its maintainability,
+future development will become possible, not only for external user datatypes,
+but also within NumPy.
+
+
+Motivation and Scope
+--------------------
+
+See also the user impact section for examples of what kind of new datatypes
+will be enabled by the proposed changes in the long run.
+
+Motivation
+^^^^^^^^^^
+
+One of the main issues with the current API is the definition of typical
+functions such as addition and multiplication for parametric datatypes
+(compare also NEP 40) require additional steps to determine the output type.
+For example when adding two strings of length four, the result is a string
+of length 8, which is different from the input.
+Similarly, a datatype which embeds a physical unit must calculate the new unit
+information: dividing a distance by a time results in a speed.
+A related difficulty is that the current casting rules
+– the conversion between different datatypes –
```

Comment:
```suggestion
---the conversion between different datatypes---
```
Possibly best to avoid Unicode and let Sphinx handle the mdash.

Reply:
Changing to `--` as en dash with spaces for now,  if unicode might make problems. I personally prefer that spelling (I know em dash without spaces is more common in current english).